base/bitarray.jl

```diff
@@ -1473,49 +1473,106 @@ maximum(B::BitArray) = isempty(B) ? error("argument must be non-empty") : any(B)
+##  q = TFTF    function  name        ##  q = TFTF    function  name        ##
+##  --------------------------------  ##  --------------------------------- ##
+##      TTTT    (true)    -           ##      FFFF    (false)   -           ##
+##      TTTF    |, max    or          ##      FFFT    ???       (nor)       ##
+##      TTFT    >=, ^     ???         ##      FFTF    <         ???         ##
+##      TTFF    (A)       -           ##      FFTT    (~A)      -           ##
+##      TFTT    <=        implies     ##      FTFF    >         notimplies  ##
+##      TFTF    (B)       -           ##      FTFT    (~B)      -           ##
+##      TFFT    ==        xnor        ##      FTTF    $, !=     xor         ##
+##      TFFF    &, *, min and         ##      FTTT    ???       (nand)      ##
+##############################################################################
+
+abstract Func{N}
+
+immutable NotFun <: Func{1} end
+call(::NotFun, x) = ~x
+
+# Use IdFun, AndFun, and OrFun from reduce.jl
+
+immutable POrNotQFun <: Func{2} end
+call(::POrNotQFun, x, y) = x | ~y
+
+immutable ImpliesFun <: Func{2} end
+call(::ImpliesFun, x, y) = ~x | y
+
+immutable XNorFun <: Func{2} end
+call(::XNorFun, x, y) = ~(x $ y)
+
+immutable NotPAndQFun <: Func{2} end
+call(::NotPAndQFun, x, y) = ~x & y
+
+immutable NotImpliesFun <: Func{2} end
+call(::NotImpliesFun, x, y) = x & ~y
+
+immutable XOrFun <: Func{2} end
+call(::XOrFun, x, y) = x $ y
+
+function specialized_unary(f::Callable)
+    is(f, !) | is(f, ~) && return NotFun()
+    is(f, identity)     && return IdFun()
+    nothing
+end
+function specialized_binary(f::Callable)
+    is(f, |)  | is(f, max) && return OrFun()
+    is(f, &)  | is(f, *)   | is(f, min) && return AndFun()
+    is(f, $)  | is(f, !=)  && return XOrFun()
+    is(f, >=) | is(f, ^)   && return POrNotQFun()
+    is(f, <=) && return ImpliesFun()
+    is(f, ==) && return XNorFun()
+    is(f, <)  && return NotPAndQFun()
+    is(f, >)  && return NotImpliesFun()
+    nothing
+end
+
+function map(f::Callable, A::BitArray)
+    t = specialized_unary(f)
+    t != nothing ? map(t, A) : invoke(map, (Callable, AbstractArray), f, A)
+end
+function map(f::Callable, A::BitArray, B::BitArray)
+    t = specialized_binary(f)
+    t != nothing ? map(t, A, B) : invoke(map, (Callable, AbstractArray, AbstractArray), f, A, B)
+end
+map(f, A::BitArray) = map!(f, similar(A), A)
+map(f, A::BitArray, B::BitArray) = map!(f, similar(A), A, B)
+
+map!(f::Callable, A::BitArray) = map!(f, A, A)
+function map!(f::Callable, dest::BitArray, A::BitArray)
+    t = specialized_unary(f)
+    t != nothing ? map!(t, dest, A) : invoke(map!, (Callable, AbstractArray, AbstractArray), f, dest, A)
+end
+function map!(f::Callable, dest::BitArray, A::BitArray, B::BitArray)
+    t = specialized_binary(f)
+    t != nothing ? map!(t, dest, A, B) : invoke(map!, (Callable, AbstractArray, AbstractArray, AbstractArray), f, dest, A, B)
+end
+
+function map!(f, dest::BitArray, A::BitArray)
```

Comment:
No, your first hunch here was right.  I initially had it more strongly typed like that but then removed it to be more like reduce.jl.